File: 17_模式匹配/isSubsequence.ts
/**
 * 判断`shorter`是否是`longer`的子序列.
 * 如果需要多次匹配，使用`子序列自动机`.
 * @complexity O(n1 + n2)
 */
function isSubsequnceNaive<S extends ArrayLike<unknown>>(longer: S, shorter: S): boolean {
  const n1 = longer.length
  const n2 = shorter.length
  if (!n2) return true
  if (n1 < n2) return false
  let i = 0
  let j = 0
  while (i < n1 && j < n2) {
    if (longer[i] === shorter[j]) j++
    if (j === n2) return true
    i++
  }
  return false
}

/**
 * 返回 {@link longer} 的每个前缀中的子序列匹配 {@link shorter} 的最大长度.
 *
 * @example
 * ```typescript
 * matchSubsequence('aabc', 'abc')
 * // [0, 1, 1, 2, 3]
 * ```
 */
function matchSubsequence<S extends ArrayLike<unknown>>(longer: S, shorter: S): number[] {
  const res = Array(longer.length + 1).fill(0)
  let i = 0
  let j = 0
  while (i < longer.length && j < shorter.length) {
    j += +(longer[i] === shorter[j])
    i++
    res[i] = j
  }
  res.fill(j, i + 1)
  return res
}

export { isSubsequnceNaive, matchSubsequence }

if (require.main === module) {
  // https://leetcode.cn/problems/is-subsequence/
  // eslint-disable-next-line no-inner-declarations
  function isSubsequence(s: string, t: string): boolean {
    return isSubsequnceNaive(t, s)
  }

  console.log(matchSubsequence('aabc', 'abc')) // [0, 1, 1, 2, 3]
}

File: 17_模式匹配/isSubsequence.go
package main

import "fmt"

func main() {
	fmt.Println(IsSubsequence([]int{1, 2, 3, 4, 5}, []int{1, 3, 5})) // true
	fmt.Println(MatchSubsequenceString("abcde", "ace"))              // [0 1 1 2 2 3]
}

// 判断`shorter`是否是`longer`的子序列.
// complexity O(n1 + n2)
func IsSubsequenceString(longer, shorter string) bool {
	if len(shorter) > len(longer) {
		return false
	}
	if len(shorter) == 0 {
		return true
	}
	n, m := len(longer), len(shorter)
	i, j := 0, 0
	for i < n && j < m {
		if longer[i] == shorter[j] {
			j++
			if j == m {
				return true
			}
		}
		i++
	}
	return false
}

// 返回`longer`的每个前缀中的子序列匹配`shorter`的最大长度.
func MatchSubsequenceString(longer, shorter string) []int {
	res := make([]int, len(longer)+1)
	i, j := 0, 0
	for i < len(longer) && j < len(shorter) {
		if longer[i] == shorter[j] {
			j++
		}
		i++
		res[i] = j
	}
	for i++; i < len(res); i++ {
		res[i] = j
	}
	return res
}

// 判断`shorter`是否是`longer`的子序列.
// 如果需要多次匹配，使用`子序列自动机`.
// complexity O(n1 + n2)
func IsSubsequence[S ~[]E, E comparable](longer, shorter S) bool {
	if len(shorter) > len(longer) {
		return false
	}
	if len(shorter) == 0 {
		return true
	}
	n, m := len(longer), len(shorter)
	i, j := 0, 0
	for i < n && j < m {
		if longer[i] == shorter[j] {
			j++
			if j == m {
				return true
			}
		}
		i++
	}
	return false
}

// 返回`longer`的每个前缀中的子序列匹配`shorter`的最大长度.
func MatchSubsequence[S ~[]E, E comparable](longer, shorter S) []int {
	res := make([]int, len(longer)+1)
	i, j := 0, 0
	for i < len(longer) && j < len(shorter) {
		if longer[i] == shorter[j] {
			j++
		}
		i++
		res[i] = j
	}
	for i++; i < len(res); i++ {
		res[i] = j
	}
	return res
}
